File: 2024/q07.jl
#!/usr/bin/env julia


const Pos = CartesianIndex{2}

lines(s) = split(s, '\n', keepempty=false)


function parseplan(text)
    d = Dict('='=>0, '+'=>1, '-'=>-1)
    s = replace(text, r"[^=+-]+"=>"")
    map(c->d[c], collect(s))
end


function parsetrack(text)
    s = stack(lines(text), dims=1)
    S = findfirst(==('S'), s)
    neibs = [(0,1), (0,-1), (1,0), (-1,0)] .|> Pos
    t, p = S, S + Pos(0, 1)
    r = s[p]
    while p != S
        for d in (p .+ neibs)
            if d != t && get(s, d, ' ') != ' '
                t, p = p, d
                r *= s[p]
                break
            end
        end
    end
    return r
end


function powerup(s::AbstractString, N::Int)
    a,s = split(s, ':')
    op = parseplan(s)
    x,r,t = 10,0,0
    for c in Iterators.cycle(op)
        (t += 1) > N && return (r, a)
        x += c
        r += x
    end
end


const _tbl = Dict((c,i)=>
        ifelse(c == 'S' || c == '=', i, c=='+' ? 1 : -1)
    for c in "S=+-" for i in [0,-1,1])

function powerup(track::AbstractString, op::AbstractVector{Int}, N::Int)
    x,r,t = 10,0,0
    n, m = length(op), length(track)
    cycl = Iterators.cycle
    if N % n == 0
        qn = lcm(n, m)
        ds = Int[]
        for p in Iterators.zip(cycl(track), cycl(op))
            push!(ds, _tbl[p])
            t += 1
            (t == qn) && break
        end
        u = gcd(N, qn) # partial run for all plans
        for _ in 1:u
            for d in ds
                x += d
                r += x
            end
        end
        return r
    else
        for p in Iterators.zip(cycl(track), cycl(op))
            x += _tbl[p]
            r += x
            (p[1] == 'S') && (t += 1)
            (t == N) && return r
        end
    end
end


function powerup(track::AbstractString, s::AbstractString, N::Int)
    a,s = split(s, ':')
    op = parseplan(s)
    r = powerup(track, op, N)
    return (r, a)
end


function part1(data, N=10)
    ns = powerup.(lines(data), N)
    join(s for (_,s) in sort(ns, rev=true))
end


function part2(track, data, N=10)
    t = parsetrack(track)
    ns = powerup.([t], lines(data), N)
    join(s for (_,s) in sort(ns, rev=true))
end


function makeplans(vs, n)
    res = Vector{Int}[]
    function inner(acc, n, a, b, c)
        if n == 0
            push!(res, acc)
            return
        end
        a > 0 && inner([acc; 1], n-1, a-1, b, c)
        b > 0 && inner([acc; 2], n-1, a, b-1, c)
        c > 0 && inner([acc; 3], n-1, a, b, c-1)
    end
    inner(Int[], n, vs...)
    return res
end


function part3(track, data, N=2024)
    t = parsetrack(track)
    score,_ = powerup(t, lines(data)[1], N)
    pl = [-1, 0, 1]
    xs = makeplans([3,3,5], 11)
    ans = 0
    for ix in xs
        s = powerup(t, pl[ix], N)
        ans += s > score
    end
    return ans
end



track1 = raw"
S+===
-   +
=+=-+
"

track2 = raw"
S-=++=-==++=++=-=+=-=+=+=--=-=++=-==++=-+=-=+=-=+=+=++=-+==++=++=-=-=--
-                                                                     -
=                                                                     =
+                                                                     +
=                                                                     +
+                                                                     =
=                                                                     =
-                                                                     -
--==++++==+=+++-=+=-=+=-+-=+-=+-=+=-=+=--=+++=++=+++==++==--=+=++==+++-
"

track3 = raw"
S+= +=-== +=++=     =+=+=--=    =-= ++=     +=-  =+=++=-+==+ =++=-=-=-- #
- + +   + =   =     =      =   == = - -     - =  =         =-=        - #
= + + +-- =-= ==-==-= --++ +  == == = +     - =  =    ==++=    =++=-=++ #
+ + + =     +         =  + + == == ++ =     = =  ==   =   = =++=        #
= = + + +== +==     =++ == =+=  =  +  +==-=++ =   =++ --= + =           #
+ ==- = + =   = =+= =   =       ++--          +     =   = = =--= ==++== #
=     ==- ==+-- = = = ++= +=--      ==+ ==--= +--+=-= ==- ==   =+=    = #
-               = = = =   +  +  ==+ = = +   =        ++    =          - #
-               = + + =   +  -  = + = = +   =        +     =          - #
--==++++==+=+++-= =-= =-+-=  =+-= =-= =--   +=++=+++==     -=+=++==+++- #
"

data = raw"
A:+,-,=,=
B:+,=,-,+
C:=,-,+,+
D:=,=,=,+
"
@assert part1(data) == "BDCA"
@assert part2(track1, data) == "DCBA"


data = readchomp("q07_p1.txt")
println("part1: ", part1(data))
data = readchomp("q07_p2.txt")
println("part2: ", part2(track2, data))
data = readchomp("q07_p3.txt")
println("part3: ", part3(track3, data))

File: event/2024/q07.py
#!/usr/bin/env python
import itertools


def part1(data, N=10):
    data = [s.split(':') for s in data.split()]
    data = [(a, s.split(',')) for a,s in data]
    m = {'+': 1, '-': -1, '=': 0}
    data = [(a, [m[x] for x in p]) for a,p in data]

    def powerup(ops):
        v = 10
        res = v
        ops = itertools.cycle(ops)
        for _ in range(N):
            v += next(ops)
            res += v
        return res

    xs = sorted((powerup(p), a) for a,p in data)
    ans = ''.join(c for _,c in xs)[::-1]
    return ans


def parse_plan(text):
    c, plan = text.split(':')
    m = {'+': 1, '-': -1, '=': 0}
    plan = tuple(m[c] for c in plan.split(','))
    return (c, plan)


def parse_track(text):
    lines = text.strip().splitlines()
    m = {'+': 1, '-': -1, '=': 0, 'S': 0}
    grid = {(x + 1j*y):c for y,row in enumerate(lines) for x,c in enumerate(row) if c != ' '}
    S = next(p for p,c in grid.items() if c == 'S')
    p = S + 1
    d = 1
    track = [m[grid[p]]]
    while p != S:
        for q in [d, d * 1j, d * -1j]:
            if (c := grid.get(p + q)):
                track.append(m[c])
                p += q
                d = q
                break
    return track


def part2(data, track, N=10):
    data = [parse_plan(s) for s in data.split()]
    track = parse_track(track)

    def powerup(ops):
        v = 10
        res = v
        ops = itertools.cycle(ops)
        for _ in range(N):
            for c in track:
                t = next(ops)
                v += c or t
                res += v
        return res

    xs = sorted((powerup(p), a) for a,p in data)
    ans = ''.join(c for _,c in xs)[::-1]
    return ans


def part3(data, track, N=2024):
    data, = [parse_plan(s) for s in data.split()]
    track = parse_track(track)

    def powerup(ops):
        v = 10
        res = v
        ops = itertools.cycle(ops)
        for _ in range(N):
            for c in track:
                t = next(ops)
                v += c or t
                res += v
        return res

    X = powerup(data[1])
    ans = 0
    seen = set()
    for p in itertools.permutations([1,1,1,1,1,-1,-1,-1,0,0,0]):
        if p in seen: continue
        seen.add(p)
        ans += powerup(p) > X
    return ans


data = """
A:+,-,=,=
B:+,=,-,+
C:=,-,+,+
D:=,=,=,+
"""
assert part1(data) == 'BDCA'

data = """
A:+,-,=,=
B:+,=,-,+
C:=,-,+,+
D:=,=,=,+
"""
track = """
S+===
-   +
=+=-+
"""
assert part2(data, track) == 'DCBA'


data = open('q07_p1.txt').read()
ans = part1(data)
print('part1:', ans)

track = """
S-=++=-==++=++=-=+=-=+=+=--=-=++=-==++=-+=-=+=-=+=+=++=-+==++=++=-=-=--
-                                                                     -
=                                                                     =
+                                                                     +
=                                                                     +
+                                                                     =
=                                                                     =
-                                                                     -
--==++++==+=+++-=+=-=+=-+-=+-=+-=+=-=+=--=+++=++=+++==++==--=+=++==+++-
"""
data = open('q07_p2.txt').read()
ans = part2(data, track)
print('part2:', ans)

track = """
S+= +=-== +=++=     =+=+=--=    =-= ++=     +=-  =+=++=-+==+ =++=-=-=--
- + +   + =   =     =      =   == = - -     - =  =         =-=        -
= + + +-- =-= ==-==-= --++ +  == == = +     - =  =    ==++=    =++=-=++
+ + + =     +         =  + + == == ++ =     = =  ==   =   = =++=
= = + + +== +==     =++ == =+=  =  +  +==-=++ =   =++ --= + =
+ ==- = + =   = =+= =   =       ++--          +     =   = = =--= ==++==
=     ==- ==+-- = = = ++= +=--      ==+ ==--= +--+=-= ==- ==   =+=    =
-               = = = =   +  +  ==+ = = +   =        ++    =          -
-               = + + =   +  -  = + = = +   =        +     =          -
--==++++==+=+++-= =-= =-+-=  =+-= =-= =--   +=++=+++==     -=+=++==+++-
"""
data = open('q07_p3.txt').read()
ans = part3(data, track)
print('part3:', ans)
